v2/api/authorization/v1api20200801preview/role_assignment_defaults.go

```diff
@@ -0,0 +1,66 @@
+// We include the namespace and name to ensure no two RoleAssignments in the same cluster can end up
+// with the same UUID.
+// We include the group and kind to ensure that different kinds of resources get different UUIDs. This isn't
+// entirely required by Azure, but it makes sense to avoid collisions between two resources of different types
+// even if they have the same namespace and name.
+// We considered also including the spec.owner.group, kind, and name here but spec.owner.name can change without
+// the actual owner AzureName changing. This could happen for example when
+// adopting the role assignment the user changed the parent Kubernetes resource name but preserved the AzureName.
+// We don't have easy access to the AzureName here so using that would be difficult.
+// Owner group + kind can't change and so could be considered in the seed string, but doing so adds additional
+// complexity to the seed algorithm. Cluster uniqueness is already guaranteed with just namespace + name.
+// Any additional seed data past namespace + name only serves to improve uniqueness between clusters (for example
+// if the user is running 3 ASO clusters). Unfortunately including owner group + kind doesn't get us much in
+// terms of guarantees as a user could still have multiple ASO instances whose RoleAssignments have the same
+// namespace/name, and are owned by the same type of resource, but whose owner has a different name and so ideally
+// the role assignments wouldn't have the same UUID.
+// The simpler namespace + name approach is easier to explain to users and works perfectly for a single instance
+// of ASO. In the rare case users have multiple ASO instances with resources in the same namespace in each cluster
+// having the same name but not actually pointing to the same Azure resource (maybe in a different subscription?)
+// they can avoid name conflicts by explicitly specifying AzureName for their RoleAssignment.
+// See https://learn.microsoft.com/en-us/azure/role-based-access-control/role-assignments#name for details about
+// RoleAssignment name restrictions. Of note is that the UUID must be unique in the AAD tenant (subscription
+// uniqueness isn't sufficient). A UUID conflict will result in the following error:
+//
+//	(RoleAssignmentUpdateNotPermitted) Tenant ID, application ID, principal ID, and scope are not allowed to be updated.
+//	Code: RoleAssignmentUpdateNotPermitted
+//	Message: Tenant ID, application ID, principal ID, and scope are not allowed to be updated.
+func (assignment *RoleAssignment) defaultAzureName() {
+	if assignment.AzureName() == "" {
+		_, err := uuid.Parse(assignment.Name)
+		if err == nil {
+			// Name is a UUID, so just use that for AzureName
+			assignment.Spec.AzureName = assignment.Name
+			return
+		}
+
+		uniqueStr := randextensions.MakeUniqueResourceString(GroupVersion.Group, fmt.Sprintf("%T", RoleAssignment{}), assignment.Namespace, assignment.Name)
```

Comment:
Why use `%T` here? Might that change between compiler releases? 
Using `RoleAssignment{}.GetType()` would return `Microsoft.Authorization/roleAssignments` which is more stable.

Reply:
> Why use %T here? Might that change between compiler releases?

It won't change at least between versions of Go 1.x due to Go's compatibility guarantees. I chose `%T` here because `RoleAssignment{}.GetType()` isn't the `Kind` and I was trying to base it on Group + Kind + Namespace + Name (the Kubernetes constructs). Your comment made me realize though that I can just do this:
```
gvk := assignment.GroupVersionKind()
uniqueStr := randextensions.MakeUniqueResourceString(gvk.Group, gvk.Kind, assignment.Namespace, assignment.Name)
```

Which is much cleaner.